Improve Regex performance (in particular RegexOptions.Compiled)

stephentoub opened the pull request:
(This builds on some experimentation that both @davidwrighton and @eerhardt had done.)

A few key changes for RegexOptions.Compiled:
- Avoid spitting out timeouts unless a timeout was actually set (the default timeout is infinite).
- Avoid calling CharInClass if the character class has a known better alternative, e.g. \d can use char.IsDigit instead.
- Avoid calling CharInClass if the character class contains a single range, e.g. [A-Z], where we can instead just emit a basic range check.
- Avoid calling CharInClass for ASCII inputs, instead using a 16-byte lookup table.  If the character class contains Unicode characters, then non-ASCII inputs fall back to still calling CharInClass.
- Avoid accessing CultureInfo.CurrentCulture unless we know we're going to need the result for ToLower calls, and we prefer to use ToLowerInvariant instead of using ToLower with CultureInfo.InvariantCulture.
- Generate more specific Ldc instructions when applicable (this is just about smaller IL).

A few general changes:
- Open-code IsWordChar and IsECMAWordChar.  We can write much more efficient implementations than using CharInClass.
- Make some internal virtual methods non-virtual, as nothing overrides them.
- Avoid some unnecessary allocations when character classes don't include categories.
- General clean-up to make working in the code easier.

A few benchmarks...

Running the RegexReduce benchmark from the dotnet/performance repo:

|         Method |        Toolchain |  options |     Mean |    Error |   StdDev |   Median |      Min |      Max | Ratio | RatioSD |      Gen 0 | Gen 1 | Gen 2 | Allocated |
|--------------- |----------------- |--------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|--------:|-----------:|------:|------:|----------:|
| RegexReduxMini | \new\corerun.exe | Compiled | 166.6 ms |  6.42 ms |  7.13 ms | 165.6 ms | 150.3 ms | 180.8 ms |  0.72 |    0.04 | 12000.0000 |     - |     - |  129.8 MB |
| RegexReduxMini | \old\corerun.exe | Compiled | 230.3 ms |  9.23 ms | 10.63 ms | 233.0 ms | 205.9 ms | 245.6 ms |  1.00 |    0.00 | 12000.0000 |     - |     - | 129.81 MB |

And running a few other random regexes:

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;
using System.Text.RegularExpressions;

[MemoryDiagnoser]
public class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssemblies(new[] { typeof(Program).Assembly }).Run(args);

    private static readonly Regex s_email = new Regex(@"^([a-zA-Z0-9_\-\.]+)@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.)|(([a-zA-Z0-9\-]+\.)+))([a-zA-Z]{2,12}|[0-9]{1,3})(\]?)$", RegexOptions.Compiled);
    private static Regex s_date = new Regex(@"\b(?<month>\d{1,2})/(?<day>\d{1,2})/(?<year>\d{2,4})\b", RegexOptions.Compiled);

    [Benchmark] public void Email() => s_email.IsMatch("yay.performance@dot.net");
    [Benchmark] public void Date() => s_date.IsMatch("Today is 11/18/2019");
}
```

| Method |        Toolchain |      Mean |    Error |   StdDev | Ratio |
|------- |----------------- |----------:|---------:|---------:|------:|
|  Email | \new\corerun.exe | 185.30 ns | 1.259 ns | 1.116 ns |  0.43 |
|  Email | \old\corerun.exe | 431.24 ns | 2.151 ns | 1.907 ns |  1.00 |
|        |                  |           |          |          |       |
|   Date | \new\corerun.exe |  97.41 ns | 3.024 ns | 3.105 ns |  0.37 |
|   Date | \old\corerun.exe | 262.80 ns | 3.529 ns | 2.947 ns |  1.00 |

cc: @danmosemsft, @adamsitnik, @ViktorHofer

stephentoub committed:
Avoid generating timeout checks for infinite timeouts

stephentoub committed:
Use Ldc_I4_X for Ldc when applicable

stephentoub committed:
Use ToLowerInvariant for invariant mode

stephentoub committed:
Remove unnecessary virtual-ness of some internal methods

stephentoub committed:
Clean up CompiledRegexRunner

stephentoub committed:
Lazily allocate RegexCharClass._categories

stephentoub committed:
Avoid negative numbers in generated names

stephentoub committed:
Optimize Is{ECMA}WordChar

stephentoub committed:
Optimize common CharInClass calls

stephentoub committed:
Clean up RegexCompiler.cs

stephentoub committed:
Add more RegexOptions.Compiled tests

stephentoub committed:
Replace RegexParser.s_category byte[] with a span

stephentoub committed:
Avoid delegate allocations each time CreateInstance is called

stephentoub committed:
Replace CharUnicodeInfo.GetUnicodeCategory w/ char's version

It has a fast path for ASCII.

stephentoub committed:
Clean up stackallocs

eerhardt reviewed (approved):
This looks great, @stephentoub! I'm seeing very impressive performance improvements with this change.

stephentoub committed:
Remove an unnecessary box

stephentoub committed:
Avoid string allocation for negated category

stephentoub committed:
Simplify tests

stephentoub committed:
Make low-hanging fruit allocation reduction

e.g. avoiding allocating a range list if a class contains only categories, avoiding some intermediary strings, avoiding some delegate allocations for sorting, etc.
